File: CocoaSlideCollection/Model/AAPLTag.m
/*
    Copyright (C) 2015 Apple Inc. All Rights Reserved.
    See LICENSE.txt for this sample’s licensing information
    
    Abstract:
    This is the "Tag" class implementation.
*/

#import "AAPLTag.h"
#import "AAPLImageFile.h"

@implementation AAPLTag

- (id)initWithName:(NSString *)newName {
    self = [super init];
    if (self) {
        name = [newName copy];
        imageFiles = [[NSMutableArray alloc] init];
    }
    return self;
}

@synthesize name;
@synthesize imageFiles;

- (void)insertImageFile:(AAPLImageFile *)imageFile {
    NSUInteger insertionIndex = [imageFiles indexOfObject:imageFile inSortedRange:NSMakeRange(0, [imageFiles count]) options:NSBinarySearchingInsertionIndex usingComparator:^NSComparisonResult(AAPLImageFile *imageFile1, AAPLImageFile *imageFile2) {
        return [imageFile1.filenameWithoutExtension caseInsensitiveCompare:imageFile2.filenameWithoutExtension];
    }];
    if (insertionIndex == NSNotFound) {
        NSLog(@"** Couldn't determine insertionIndex for imageFiles array");
    } else {
        [imageFiles insertObject:imageFile atIndex:insertionIndex];
    }
}

- (NSString *)description {
    return [NSString stringWithFormat:@"{Tag: %@}", self.name];
}

@end

File: CocoaSlideCollection/Model/AAPLTag.swift
//
//  AAPLTag.swift
//  CocoaSlideCollection
//
//  Translated by OOPer in cooperation with shlab.jp, on 2015/12/23.
//
//
/*
    Copyright (C) 2015 Apple Inc. All Rights Reserved.
    See LICENSE.txt for this sample’s licensing information

    Abstract:
    This is the "Tag" class declaration.
*/

import Cocoa

// An AAPLTag is a label string that can be applied to ImageFiles.  An AAPLImageCollection has a list of Tags, each of which has associated ImageFiles.
@objc(AAPLTag)
class AAPLTag: NSObject {
    private(set) var name: String                 // the tag string (e.g. "Vacation")
    @objc dynamic private(set) var imageFiles: [AAPLImageFile] = []     // the ImageFiles that have this tag, ordered for display using our desired sort
    
    init(name newName: String) {
        name = newName
        super.init()
    }
    
    func insertImageFile(_ imageFile: AAPLImageFile) {
        let insertionIndex = imageFiles.indexOf(imageFile, inSortedRange: imageFiles.startIndex..<imageFiles.endIndex) {imageFile1, imageFile2 in
            return imageFile1.filenameWithoutExtension!.caseInsensitiveCompare(imageFile2.filenameWithoutExtension!)
        }
        imageFiles.insert(imageFile, at: insertionIndex)
    }
    
    override var description: String {
        return "{Tag: \(self.name)}"
    }
    
}
